### PR TITLE
Libsnark get_evaluation_domain works on macOS

### PR DESCRIPTION
**Pull request by bkase**
Before this change on macOS systems, libsnark would crash on
get_root_of_unity and exception handlers would not work. The
hypothesis is that using exceptions for control flow in
get_evaluation_domain triggers some sort of UB (probably because this
runs statically before the main function). On Linux, the UB happens to
work properly, but on OSX it doesn't. As such, I rewrote all the
control-flow related exception handling in get_evaluation_domain to use
an error bool pointer.

I'll open a PR against libsnark directly when the review for this PR
finishes.

On macOS the Coda build succeeds only after these changes are made (as
we use libsnark at compile time). CI will tell us that the build still
works on Linux.

### CODE REVIEW

File: src/lib/snarky/src/camlsnark_c/libsnark-caml/depends/libff/libff/algebra/fields/field_utils.tcc [hunk 1]
```diff
@@ -16,6 +16,11 @@
 #include <libff/common/double.hpp>
 #include <libff/common/utils.hpp>
 
+#include <execinfo.h>
```
Comment (cmr) [inline src/lib/snarky/src/camlsnark_c/libsnark-caml/depends/libff/libff/algebra/fields/field_utils.tcc]: unneeded

**Comment by cmr** [comment]
`bool*` should be `bool&`

**Comment by cmr** [comment]
There are ways to avoid initializing those fields on exit paths, but the code is already there, and it's the FieldTs that are expensive anyway. I'm fine with it. Thank you for not letting uninitialized data escape.

**Comment by HarryR** [comment]
Dredging this back up as it's relevant again for me.

> The hypothesis is that using exceptions for control flow in...
Is the problem.

It's preventing two things:

 1) Optimized builds on OSX (enabling `-O3` triggers this, normal builds on OSX work fine)
 2) [Cheerp](https://www.leaningtech.com/cheerp/) WASM build, which are unable to catch exceptions so `abort()` instead.

It seems no pull requests are getting merged into libsnark, so will have to fork and apply this to my branch.

**Pull request merged**